torch/distributed/_spmd/api.py

```diff
@@ -314,6 +382,12 @@ def stateless_func(
                     _allow_non_fake_inputs=False,
                 )(args, kwargs, named_states, params_and_buffers)
 
+            # FIXME(@mrshenli): functionalization does not work for our use case
+            # yet. I so implemented a mini-functionalization just for foreach
+            # ops. Remove this when the following issue is addressed.
+            # Issue: https://github.com/pytorch/pytorch/issues/97852
+            gm = _functionalize_and_reinplace_foreach_ops(gm)
```

Comment:
This replaces `foreach_add_()` with `foreach_add()` + `copy_()`, right?

Are you still running into the proxy_tensor.py issue that you mentioned earlier, the first time that `foreach_add_()` is traced?

Reply:
Hey @bdhirsh 

> This replaces foreach_add_() with foreach_add() + copy_(), right?

Yep, exactly. Below is the graph after `_functionalize_and_reinplace_foreach_ops `

```
opcode         name               target                                args                                      kwargs                                          
-------------  -----------------  ------------------------------------  ----------------------------------------  ------------------------------------------------
-------------                                                                                                                                                     
placeholder    arg0_1             arg0_1                                ()                                        {}                                              
placeholder    arg0_2             arg0_2                                ()                                        {}                                              
placeholder    arg0_3             arg0_3                                ()                                        {}                                              
placeholder    arg2_1             arg2_1                                ()                                        {}                                              
placeholder    arg2_2             arg2_2                                ()                                        {}                                              
placeholder    arg3_1             arg3_1                                ()                                        {}                                              
placeholder    arg3_2             arg3_2                                ()                                        {}                                              
call_function  t                  aten.t.default                        (arg3_1,)                                 {}                                              
call_function  addmm              aten.addmm.default                    (arg3_2, arg0_3, t)                       {}                                              
call_function  sum_1              aten.sum.default                      (addmm,)                                  {}                                              
call_function  ones_like          aten.ones_like.default                (sum_1,)                                  {'pin_memory': False, 'memory_format': torch.pre
serve_format}                                                                                                                                                     
call_function  sym_size           aten.sym_size                         (arg0_3, 0)                               {}                                              
call_function  sym_size_1         aten.sym_size                         (arg3_1, 0)                               {}                                              
call_function  expand             aten.expand.default                   (ones_like, [sym_size, sym_size_1])       {}                                              
call_function  t_1                aten.t.default                        (expand,)                                 {}                                              
call_function  mm                 aten.mm.default                       (t_1, arg0_3)                             {}                                              
call_function  t_2                aten.t.default                        (mm,)                                     {}                                              
call_function  sum_2              aten.sum.dim_IntList                  (expand, [0], True)                       {}                                              
call_function  sym_size_2         aten.sym_size                         (arg3_2, 0)                               {}                                              
call_function  view               aten.view.default                     (sum_2, [sym_size_2])                     {}                                              
call_function  detach             aten.detach.default                   (view,)                                   {}
call_function  detach_1           aten.detach.default                   (detach,)                                 {}
call_function  t_3                aten.t.default                        (t_2,)                                    {}
call_function  detach_2           aten.detach.default                   (t_3,)                                    {}
call_function  detach_3           aten.detach.default                   (detach_2,)                               {}
call_function  _foreach_add_list  aten._foreach_add.List                ([arg3_1, arg3_2], [detach_3, detach_1])  {'alpha': -0.01}
call_function  getitem_1          <function getitem at 0x7f10cdeb3f70>  (_foreach_add_list, 1)                    {}
call_function  copy__1            aten.copy_                            (arg3_2, getitem_1)                       {}
call_function  getitem            <function getitem at 0x7f10cdeb3f70>  (_foreach_add_list, 0)                    {}
call_function  copy_              aten.copy_                            (arg3_1, getitem)                         {}
output         output             output                                ([None, copy_, copy__1],)                 {}
```

Reply:
> Are you still running into the proxy_tensor.py issue that you mentioned earlier, the first time that foreach_add_() is traced?

`_functionalize_and_reinplace_foreach_ops` kind of walk around this problem. Just to be clear, the original problem was that the output of `aten._foreach_add_.List` (which None) was not connected to graph output, as a result, we cannot figure out dependencies correctly. 

```
opcode         name           target                   args                                      kwargs                                                  
-------------  -------------  -----------------------  ----------------------------------------  -------------------------------------------------------------    
placeholder    arg0_1         arg0_1                   ()                                        {}                                                               
placeholder    arg0_2         arg0_2                   ()                                        {}                                                               
placeholder    arg0_3         arg0_3                   ()                                        {}                                                               
placeholder    arg2_1         arg2_1                   ()                                        {}                                                               
placeholder    arg2_2         arg2_2                   ()                                        {}                                                               
placeholder    arg3_1         arg3_1                   ()                                        {}                                                               
placeholder    arg3_2         arg3_2                   ()                                        {}                                                               
call_function  t              aten.t.default           (arg3_1,)                                 {}                                                               
call_function  addmm          aten.addmm.default       (arg3_2, arg0_3, t)                       {}                                                               
call_function  sum_1          aten.sum.default         (addmm,)                                  {}                                                               
call_function  ones_like      aten.ones_like.default   (sum_1,)                                  {'pin_memory': False, 'memory_format': torch.preserve_format}    
call_function  sym_size       aten.sym_size            (arg0_3, 0)                               {}                                                               
call_function  sym_size_1     aten.sym_size            (arg3_1, 0)                               {}                                                               
call_function  expand         aten.expand.default      (ones_like, [sym_size, sym_size_1])       {}                                                               
call_function  t_1            aten.t.default           (expand,)                                 {}                                                               
call_function  mm             aten.mm.default          (t_1, arg0_3)                             {}                                                               
call_function  t_2            aten.t.default           (mm,)                                     {}                                                               
call_function  sum_2          aten.sum.dim_IntList     (expand, [0], True)                       {}                                                               
call_function  sym_size_2     aten.sym_size            (arg3_2, 0)                               {}                                                               
call_function  view           aten.view.default        (sum_2, [sym_size_2])                     {}                                                               
call_function  detach         aten.detach.default      (view,)                                   {}                                                               
call_function  detach_1       aten.detach.default      (detach,)                                 {}                                                               
call_function  t_3            aten.t.default           (t_2,)                                    {}                                                               
call_function  detach_2       aten.detach.default      (t_3,)                                    {}                                                               
call_function  detach_3       aten.detach.default      (detach_2,)                               {}                                                               
call_function  _foreach_add_  aten._foreach_add_.List  ([arg3_1, arg3_2], [detach_3, detach_1])  {'alpha': -0.01}                                                 
output         output         output                   ([None, arg3_1, arg3_2],)                 {}   
```

Reply:
I think there might actually be a more elegant solution here than writing this graph pass: when you run make_fx(), you can specify a decomposition table - so one option is just to include a decomposition, from `_foreach_add_() into _foreach_add() + copy_()!

```
def foreach_add_decomp(self, other, scalar=1):
    self_updated = torch.ops.aten._foreach_add.List(self, other, scalar=scalar)
    for s, s_u in zip(self, self_updated):
        s.copy_(s_u)

...

gm = make_fx(
                    partial(stateless_func, func),
                    tracing_mode="symbolic",
                    decomposition_table={torch.ops.aten._foreach_add_.List: foreach_add_decomp},
                    _allow_non_fake_inputs=False,
                )(args, kwargs, named_states, params_and_buffers)
```

Reply:
wow, this is amazing! Let me try this. Thanks!

Reply:
> _functionalize_and_reinplace_foreach_ops kind of walk around this problem.

I only mention that because I see that you're running make_fx() first to get a graph with `_foreach_add_()`, and then you're manually running the graph pass afterwards. I thought you might be hitting the problem you mentioned in this bit of code, with proxy tensor (but I could be wrong): https://github.com/pytorch/pytorch/blob/2c16b73a1b310a58a0a1e8544cb60c133e753d6f/torch/fx/experimental/proxy_tensor.py#L359

Reply:
my bad, I didn't register meta ops correctly